src/mlpack/methods/ann/gan.hpp

```diff
@@ -0,0 +1,187 @@
+
+  /**
+   * Gradient function for gan. 
+   * This function is passes the gradient based
+   * on which network is being trained ie generator or Discriminator.
+   * 
+   * @param parameters present parameters of the network
+   * @param i index of the predictors
+   * @param gradient variable to store the present gradient
+   */
+  void Gradient(const arma::mat& parameters, const size_t i,
+      arma::mat& gradient);
+
+  /**
+   * This function does forward pass through the GAN
+   * network.
+   *
+   * @param input Sampled noise
+   */
+  void Forward(arma::mat&& input);
+
+  /**
+   * This function predicts the output of the network
+   * on the given input.
+   *
+   * @param input  the input  the discriminator network
+   * @param output result of the discriminator network
+   */
+  void Predict(arma::mat&& input, arma::mat& output);
+
+  //! Return the parameters of the network.
+  const arma::mat& Parameters() const { return parameter; }
+  //! Modify the parameters of the network
+  arma::mat& Parameters() { return parameter; }
+
+  //! Return the number of separable functions (the number of predictor points).
+  size_t NumFunctions() const { return numFunctions; }
+
+  //! Serialize the model.
+  template<typename Archive>
+  void Serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  //! Locally stored parameter for training data.
+  arma::mat predictors;
+  //! Locally stored parameters of the network.
+  arma::mat parameter;
+
+  //! Locally stored generator network.
+  Model& generator;
+  //! Locally stored discriminator network.
+  Model& discriminator;
+  //! Locally stored Intialiser.
+  InitializationRuleType  initializeRule;
+  //! Locally stored Noise function
+  Noise noiseFunction;
+  size_t noiseDim;
+  //! Locally stored number of data points.
+  size_t numFunctions;
+
+  //! Locally stored batch size parameter.
+  size_t batchSize;
+
+  //! Locally stored offset for predictors and noise data.
+  size_t offset;
+  //! Locally stored number of iterations that have been completed.
+  size_t counter;
+  //! Locally stored batch number which is being processed.
+  size_t currentBatch;
+
+  size_t generatorUpdateStep;
+
+  size_t preTrainSize;
+
+  //! Locally stored reset parmaeter.
+  bool reset;
+  //! Locally stored delta visitor.
+  DeltaVisitor deltaVisitor;
+  //! Locally stored responses.
+  arma::mat responses;
+  //! Locally stored current input.
+  arma::mat currentInput;
+  //! Locally stored current target.
+  arma::mat currentTarget;
+  //! Locally-stored output parameter visitor.
+  OutputParameterVisitor outputParameterVisitor;
+  //! Locally-stored weight size visitor.
+  WeightSizeVisitor weightSizeVisitor;
+  //! Locally-stored reset visitor.
+  ResetVisitor resetVisitor;
+  //! Locally stored gradient parameters.
+  arma::mat gradient;
+  //! Locally stored gradient for discriminator.
+  arma::mat gradientDiscriminator;
+
+  arma::mat noiseGradientDiscriminator;
+
+  arma::mat noise;
```

Comment:
It seems the variable is not used.

Reply:
It is still used. Look at the evaluate function and the gradient function.